Add entity groups

ryan-carroll-graylog opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

kingzacko1 commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/EntityType.java:
```diff
@@ -0,0 +1,36 @@
+    @JsonProperty("assets")
+    ASSETS,
+    @JsonProperty("sigma_rules")
+    SIGMA_RULES;
```
Since this is in open we may want to find a way to keep enterprise entity references out as much as possible. Might be okay to leave them as strings and then from the entity service classes we can use the corresponding `EntityDBService.COLLECTION_NAME` variable to consistently set/retrieve the values.

ryan-carroll-graylog replied:
Yeah good point. This was merely forward looking to how we'll handle mapping the different types to their respective content pack entity type handling. But I know we have other ways of doing that so I think we can get a away with just String type keys, and will dig in more once we get to the content pack piece.

danotorrey commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java:
```diff
@@ -0,0 +1,143 @@
+        if (EntityGroup.id() != null) {
+            return scopedEntityMongoUtils.update(EntityGroup);
+        }
+        String newId = scopedEntityMongoUtils.create(EntityGroup);
```
When trying to update an entity group (without supplying an ID in the request JSON), the update call ends up creating a new copy of the entity. I think some other recently developed API requests don't rely on the ID being in the JSON but instead use the ID from the URL to look up the existing entity (and error if it's not found). I think doing the same here would be good.

ryan-carroll-graylog replied:
Ah yeah nice catch, just updated to use the url ID.

danotorrey commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java:
```diff
@@ -0,0 +1,143 @@
+                paginationHelper.filter(searchQuery.toBson()).sort(sort).perPage(perPage).page(page, filter);
+    }
+
+    public EntityGroup save(EntityGroup EntityGroup) {
```
```suggestion
    public EntityGroup save(EntityGroup entityGroup) {
```
Camel case

ryan-carroll-graylog replied:
Thanks for the catches! Looks like some bad find-and-replaces from the original categories verbiage.

danotorrey commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/BulkEntityGroupRequest.java:
```diff
@@ -0,0 +1,31 @@
+        List<String> entityIds
+) {
+    public static final String FIELD_ENTITY_TYPE = "entity_type";
+    public static final String FIELD_ENTITY_IDS = "entity_IDS";
```
```suggestion
    public static final String FIELD_ENTITY_IDS = "entity_ids";
```
Nit: lower case

ryan-carroll-graylog replied:
Thanks, just updated

danotorrey commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupPermissions.java:
```diff
@@ -0,0 +1,22 @@
+package org.graylog2.entitygroups.rest;
+
+public class EntityGroupPermissions {
+    public static final String ENTITY_GROUP_READ = "category:read";
+    public static final String ENTITY_GROUP_EDIT = "category:edit";
```
Would we want to use the `entity_group` name in the permissions and audit names for consistency, or were we planning to refer to them as categories in the frontend?

danotorrey replied:
Not sure if we need to go there now, but it might be helpful to have a `type` enum for entity groups, so that they could be used for specific use cases. For example with Categories, we could assign a `CATEGORY` type, as a sort of namespace for all the security-specific categories. Not sure it's worth the effort at this point, but might be useful to nicely solve the naming difference.

danotorrey replied:
Makes sense to me, definitely good to wait until we have a use case.

danotorrey reviewed (changes requested):
@ryan-carroll-graylog This is looking good to me and also tests successfully! Just noticed one issue with the update endpoint.

danotorrey reviewed (approved):
LGTM and tests successfully with the latest!

kingzacko1 commented:
@bernd this is the backend groundwork of the Entity Groups (name we've landed on for the Content Tagging/Categorization feature). Idea is for the DTO to look something like (with empty arrays omitted): 

```json
[
	{
		"_id": "661d3e54742f917acb3c675b",
		"_scope": "ILLUMINATE",
		"name": "Windows Compliance",
		"entities": {
			"event_definitions": [],
			"sigma_rules": ["661d3e52742f917acb3c675a", "661d3e52742f917acb3c675b", "661d3e52742f917acb3c675c"],
			"anomaly_detectors": [],
			"dashboards": [ "661d3e54742f917acb3c675a", "661d3e54742f917acb3c674c", "661d3e54742f917acb3c6888" ]
			"saved_searches": [],
			"streams": [ "651d3e54742f917acb3c675a" ],
			"content_packs": []
		}
	},
	{
		"_id": "661d3e54742f917acb3c671f",
		"_scope": "DEFAULT",
		"name": "User's Custom Group",
		"entities": {
			"event_definitions": [],
			"sigma_rules": [],
			"anomaly_detectors": [],
			"dashboards": [ "661d3eaaaa2f917acb3c675a", "661d3eaaaa2f917acb3c674c", "661d3eaaaa2f917acb3c6888" ]
			"saved_searches": [],
			"streams": [],
			"content_packs": ["661d3eaaaa2f917acb3c6fec", "661d3eaaaa2f917acbacd75b"]
		}
	}
]
```

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java:
```diff
@@ -0,0 +1,143 @@
+                in(typeField(type), entityIds)
+        );
+        final List<EntityGroup> groups = MongoUtils.stream(collection.find(query)).toList();
+        final Multimap<String, EntityGroup> entityToGroupsMap = MultimapBuilder.hashKeys().hashSetValues().build();
+        for (EntityGroup group : groups) {
+            for (Map.Entry<String, Set<String>> StringEntry : group.entities().entrySet()) {
+                if (StringEntry.getKey().equals(type)) {
+                    for (String entityId : StringEntry.getValue()) {
+                        if (entityIds.contains(entityId)) {
+                            entityToGroupsMap.put(entityId, group);
+                        }
+                    }
+                }
+            }
+        }
```
Some comments:
* `group.entities()` is nullable, so it should be checked, however in practice the query guarantees that the group will have entities. If that's always the case on the DB level, maybe reconsider the `@Nullable` annotation on `entities()`.
* Loop over `entities().entrySet()` and `if (StringEntry.getKey().equals(type)) {` should be a simple `get(type)`? The query already makes sure all groups contain something from that type, we just don't know which entityId it is.
* The overall code would benefit from performing this as an aggregation query in MongoDB, instead of doing it in code.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java:
```diff
@@ -0,0 +1,143 @@
+                }
+            }
+        }
+        return entityToGroupsMap.asMap();
```
If this is a `MultiMap` why not return that type and use it consistently?

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java:
```diff
@@ -0,0 +1,143 @@
+                exists(typeField(type)),
+                in(typeField(type), entityId)
+        );
+        return MongoUtils.stream(collection.find(query)).toList();
```
As a general rule, please allow the caller to make the decision whether they want a stream or a list. `toList` in database services are usually not the right approach.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/EntityGroup.java:
```diff
@@ -0,0 +1,97 @@
+    @JsonProperty(FIELD_NAME)
+    public abstract String name();
+
+    @Nullable
```
As noted in the comment for `DBEntityGroupService`, please reconsider whether this should be nullable or default to an empty set.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+    @Path("/{id}")
+    @ApiOperation(value = "Get a single entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_READ)
```
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+    @GET
+    @Path("/get_for_entity")
+    @ApiOperation(value = "Get a list of entity groups for an entity")
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_READ)
```
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Create a new entity group")
+    @ApiResponses(@ApiResponse(code = 400, message = "An entity group already exists with id or name"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
```
This is fine as an annotation, because there is no instance to check yet, but this should use a "create" permission instead of "edit".

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+            @ApiResponse(code = 400, message = "An entity group already exists with id or name"),
+            @ApiResponse(code = 404, message = "No such entity group")
+    })
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
```
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+    @Path("/{group_id}/add_entity")
+    @ApiOperation("Add an entity to an entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
```
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+    @PUT
+    @Path("/get_for_entities")
+    @ApiOperation("Get a list of entity groups for a list of entities")
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_READ)
```
Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,178 @@
+    @Path("/{id}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Delete an entity group")
+    @RequiresPermissions(EntityGroupPermissions.ENTITY_GROUP_EDIT)
```
Like the others, but this also should use a separate "delete" permission.

Permission checks on resources that deal with a specific instance should use per-instance checks, not only annotations. Compare eg. https://github.com/Graylog2/graylog2-server/blob/e53295c35322939ad25a4d3eb67196604bd5d321/graylog2-server/src/main/java/org/graylog2/rest/resources/users/UsersResource.java#L191

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupPermissions.java:
```diff
@@ -0,0 +1,22 @@
+ */
+package org.graylog2.entitygroups.rest;
+
+public class EntityGroupPermissions {
```
This is missing both delete and create

kroepke reviewed (changes requested):
I've left a few comments inline.
High-level comments:
* we are still missing tests for the logic (no need to test infrastructure of course)
* please have a look at using document update query commands from MongoDB instead of load-modify-store patterns, as it is inherently dangerous and prone to inconsistent results (and also a performance problem at scale)

ryan-carroll-graylog commented:
Really appreciate the review and feedback @kroepke! I believe I've addressed everything you noted but definitely let me know if I missed anything.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/model/DBEntityGroupService.java:
```diff
@@ -0,0 +1,164 @@
+                in(typeField(type), entityIds)
+        );
+        final Multimap<String, EntityGroup> entityToGroupsMap = MultimapBuilder.hashKeys().hashSetValues().build();
+        MongoUtils.stream(collection.find(query)).forEach(group -> {
```
Stream uses must call close, because this is an IO stream (see the javadocs of MongoUtils#stream).
`.forEach` does not call `close()`, although one could assume it does.

The best way to do this is always to use try-with-resources on stream usages; worst case, it's a no-op for streams that don't require it:

```java
try (final Stream<EntityGroup> stream = MongoUtils.stream(collection.find(query))) {
            stream.forEach(group -> { ... });
}
```

ryan-carroll-graylog replied:
Wow I was not aware of this, thanks for pointing that out! I've addressed it here and will work this pattern in elsewhere as  I see it.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+    @Path("/{id}")
+    @ApiOperation(value = "Get a single entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(ENTITY_GROUP_READ)
```
You shouldn't use the annotation if you check the individual permission in the code as that might confuse future readers.
Another PR will add marker annotations to clearly spell out the intent, but for now, let's use either an annotation or the code check, but not both at the same time (also reduces the risk of modifying one but not the other).

ryan-carroll-graylog replied:
Ok definitely makes sense, have removed the extraneous annotations in this class.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+    @GET
+    @Path("/get_for_entity")
+    @ApiOperation(value = "Get a list of entity groups for an entity")
+    @RequiresPermissions(ENTITY_GROUP_READ)
```
Extra annotation

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+            @ApiResponse(code = 400, message = "An entity group already exists with id or name"),
+            @ApiResponse(code = 404, message = "No such entity group")
+    })
+    @RequiresPermissions(ENTITY_GROUP_EDIT)
```
Extra annotation

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+    @Path("/{group_id}/add_entity")
+    @ApiOperation("Add an entity to an entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(ENTITY_GROUP_EDIT)
```
Extra annotation

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+    @Path("/{id}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Delete an entity group")
+    @RequiresPermissions(ENTITY_GROUP_DELETE)
```
Extra annotation

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+
+    @GET
+    @ApiOperation(value = "Get a list of entity groups")
+    @RequiresPermissions(ENTITY_GROUP_READ)
```
In resources that return a list of entities to a user, the entities need to be permission-checked before being returned to the client.

While this creates an awkward problem with pagination and limits, it's currently the only way to ensure we don't leak information.

One way to do this here would be:

```java
final PaginatedList<EntityGroup> paginated = entityGroupService.findPaginated(query, page, perPage, order, sort, null);
return paginated.withList(paginated.stream().filter(entityGroup -> isPermitted(ENTITY_GROUP_READ, entityGroup.id())).toList());
```

This approach isn't great, as it could have completely empty pages or pages with less than `per_page` entries, but at least it would be correct.
There's code in the search classes that extends pagination with a predicted, but for now there's no globally available method to do this properly.

kroepke replied:
In fact there's https://github.com/Graylog2/graylog2-server/blob/1967c43ac4e4e2c9a4ab37517c9cb7cb4fffe3fb/graylog2-server/src/main/java/org/graylog2/database/pagination/MongoPaginationHelper.java#L104 which can help do this already

ryan-carroll-graylog replied:
Definitely makes sense, went ahead and refactored the paginated endpoints to use the MongoPaginationHelper filters downstream.

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+                                                 @QueryParam("sort") String sort,
+                                                 @ApiParam(name = "direction", value = "The sort direction", allowableValues = "asc,desc")
+                                                 @DefaultValue("asc") @QueryParam("direction") SortOrder order) {
+        return entityGroupService.findPaginated(query, page, perPage, order, sort, null);
```
```suggestion
final PaginatedList<EntityGroup> paginated = entityGroupService.findPaginated(query, page, perPage, order, sort, null);
return paginated.withList(paginated.stream().filter(entityGroup -> isPermitted(ENTITY_GROUP_READ, entityGroup.id())).toList());
```

kroepke commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,207 @@
+    @PUT
+    @Path("/get_for_entities")
+    @ApiOperation("Get a list of entity groups for a list of entities")
+    @RequiresPermissions(ENTITY_GROUP_READ)
```
same as with findPaginated, this needs to check the returned entities individually

ryan-carroll-graylog commented:
Thanks for all the comments @kroepke! A lot of great info I believe I've addressed here, and will apply across the board elsewhere too.

danotorrey commented on graylog2-server/src/main/java/org/graylog2/audit/AuditEventTypes.java:
```diff
@@ -172,6 +172,10 @@ public class AuditEventTypes implements PluginAuditEventTypes {
     public static final String TELEMETRY_USER_SETTINGS_UPDATE = PREFIX + "telemetry_user_settings:update";
     public static final String CONTENT_STREAM_USER_SETTINGS_UPDATE = PREFIX + "content_stream_user_settings:update";
 
+    public static final String ENTITY_GROUP_CREATE = PREFIX + "category:create";
+    public static final String ENTITY_GROUP_UPDATE = PREFIX + "category:update";
+    public static final String ENTITY_GROUP_DELETE = PREFIX + "category:delete";
```
I think this might have been missed in the recent changes, but I believe we were going to use the name `entity_group` everywhere.
```suggestion
    public static final String ENTITY_GROUP_CREATE = PREFIX + "entity_group:create";
    public static final String ENTITY_GROUP_UPDATE = PREFIX + "entity_group:update";
    public static final String ENTITY_GROUP_DELETE = PREFIX + "entity_group:delete";
```

danotorrey commented on graylog2-server/src/main/java/org/graylog2/audit/AuditEventTypes.java:
```diff
@@ -319,6 +323,9 @@ public class AuditEventTypes implements PluginAuditEventTypes {
             .add(TELEMETRY_USER_SETTINGS_UPDATE)
             .add(CONTENT_STREAM_USER_SETTINGS_UPDATE)
             .add(CERTIFICATE_RENEWAL_MANUALLY_INITIATED)
+            .add(ENTITY_GROUP_CREATE)
+            .add(ENTITY_GROUP_UPDATE)
+            .add(ENTITY_GROUP_DELETE)
```
It looks like formatters for these audit entries are still needed, since the general message is still being logged.
<img width="905" alt="image" src="https://github.com/Graylog2/graylog2-server/assets/3423655/fad07a19-e319-4253-adb6-eca192e83c1b">

danotorrey commented on graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java:
```diff
@@ -0,0 +1,202 @@
+                                                 @ApiParam(name = "query") @QueryParam("query") @DefaultValue("") String query,
+                                                 @ApiParam(name = "sort",
+                                                           value = "The field to sort the result on",
+                                                           allowableValues = "name")
```
Nit: Similar to some other pagination endpoints, could we add a default value in the pagination endpoints to avoid an error when it's not provided? 
```suggestion
                                                           allowableValues = "name")
                                                 @DefaultValue(EntityGroup.FIELD_NAME)
```

danotorrey reviewed (changes requested):
@ryan-carroll-graylog Thanks for all of the updates. I just noticed a couple of additional items from the latest.

ryan-carroll-graylog commented:
Thanks for reviewing with the latest @danotorrey! Somehow I missed the `categories` audit logs last time around.
Pushed some updates here and opened an enterprise PR for the audit log formatting: https://github.com/Graylog2/graylog-plugin-enterprise/pull/7585

ryan-carroll-graylog commented:
@kroepke I believed I've addressed all the PR feedback here. Do you see any issues merging this once we get the last SecDev reviews so we can push forward with front end development? 

We've got this behind a feature flag so we can also safely make changes in the future as needed without any major disruptions.

danotorrey reviewed (approved):
Thanks for all the fixes @ryan-carroll-graylog! LGTM and tests successfully.

danotorrey commented:
We discussed this with the Architecture team, and identified that we are putting the general Entity Groups solution on hold. We may pursue it further later, as a more general way to link entities of different types into a group with potential additional properties such as permissions etc.

In the short-term, we will pursue https://github.com/Graylog2/graylog-plugin-enterprise/issues/7926 instead.

bernd commented:
> We discussed this with the Architecture team, and identified that we are putting the general Entity Groups solution on hold. We may pursue it further later, as a more general way to link entities of different types into a group with potential additional properties such as permissions etc.
> 
> In the short-term, we will pursue [Graylog2/graylog-plugin-enterprise#7926](https://github.com/Graylog2/graylog-plugin-enterprise/issues/7926) instead.

I am changing the PR to draft status.
